### ISSUE TITLE
Cannot use target_compatible_with to ignore MacOS only targets

### ISSUE THREAD

**Issue by jagobagascon** [bazelbuild/bazel#12897]
### Description of the problem:

I'm trying to use the `target_compatible_with` attribute to automatically skip MacOS only targets on other platforms, but I just get an error every time. I have a simple `objc_library` rule with `target_compatible_with = [ "@platforms//os:macos" ]` but whenever I run `bazel build //...:all` on non MacOS platforms I get this:

```log
ERROR: /root/.cache/bazel/_bazel_root/eab0d61a99b6696edb3d2aff87b585e8/external/local_config_cc/BUILD:47:19: in cc_toolchain_suite rule @local_config_cc//:toolchain: cc_toolchain_suite '@local_config_cc//:toolchain' does not contain a toolchain for cpu 'darwin_x86_64'
ERROR: Analysis of target '//ble/platform/macos:ble_macos' failed; build aborted: Analysis of target '@local_config_cc//:toolchain' failed
```

### Bugs: what's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

Any `objc_library` rule fails instead of being ignored.
```bazel
objc_library(
    name = 'ble',
    target_compatible_with = [
        "@platforms//os:macos",
    ],
)
```

I added a simple example here: [jagobagascon/bazel-target-compatible](https://github.com/jagobagascon/bazel-target-compatible)
It's just an empty `objc_library` rule but you can se how the Ubuntu build fails while the MacOS just works: [jagobagascon/bazel-target-compatible/actions/runs/512144112](https://github.com/jagobagascon/bazel-target-compatible/actions/runs/512144112)

### What operating system are you running Bazel on?

I'm running it on a Docker image based on `ubuntu:focal-20200423`, but it also happens on a GitHub-hosted `ubuntu-18.04`.

### What's the output of `bazel info release`?

release 4.0.0

**Comment by katre** [comment]
I can reproduce this: it looks like [`objc_library`](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/rules/objc/ObjcLibraryRule.java?q=ObjcLibraryRule&ss=bazel) has a self-transition ([`AppleCrosstoolTransition`](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/rules/objc/AppleCrosstoolTransition.java;drc=37ee252f3744abc4511f55b5089cc52abd3ba09d;l=65)), which changes the value of the `--cpu` flag to one that's appropriate for building objc code. Unfortunately, this causes two issues:

1.  You don't have a cc toolchain for the selected cpu value (`darwin_x86_64`, looks like), which causes an analysis failure.
2.  This failure happens _before_ the incompatible target skipping code runs.

I'm passing this off the @gregestren to see if there's a way to restructure the target skipping to ignore analysis failures in dependencies, but I'm not sure this is possible.

**Comment by katre** [comment]
The target platform is loaded before toolchains are actually resolved, but I'm not sure there's a good way to pass that information back out to ConfiguredTargetFunction along with the resolution error.

There are a lot of arguments for making all toolchains optional (and letting rules themselves throw errors if they need a toolchain that doesn't exist), and this seems to be one of them.

**Comment by gregestren** [comment]
Do you think `ConfiguredTargetFunction` could just call https://github.com/bazelbuild/bazel/blob/8346ea4cfdd9fbd170d51a528fee26f912dad2d5/src/main/java/com/google/devtools/build/lib/skyframe/PlatformLookupUtil.java#L50 directly, constructing that platform key with https://github.com/bazelbuild/bazel/blob/8346ea4cfdd9fbd170d51a528fee26f912dad2d5/src/main/java/com/google/devtools/build/lib/skyframe/ToolchainResolutionFunction.java#L212-L216?

If so, the other challenge would be to short-circuit return a `ConfiguredTarget` hard-set to `IncompatiblePlatformProvider`. I think that'd be straightforward, albeit it complicates `ConfiguredTargetFunction` logic even more.

Re: generalizing optional toolchains, @katre offered to write up deeper thoughts on that.

CC @philsc for FYI, or if I'm getting anything wrong.

**Comment by katre** [comment]
I think that would work to get a target platform. I can't say about the logic for the incompatible platform provider.

**Comment by gregestren** [comment]
To be clear, I'd really like to prototype this but I don't have time at the moment. I am keeping this issue in my back burner memory - I know it's important to work out first bugs on a new feature to really make the feature land well.

**Comment by philsc** [comment]
I'll take a stab at your proposed idea this weekend and report back with what I find. It sounds reasonable, but also sounds like it could mean moving the entire logic to a completely different spot.

**Comment by gregestren** [comment]
Awesome, thanks! I think if injected carefully the change won't be overly invasive. But yeah it's a bit delicate to nudge all the parts exactly right.

**Comment by philsc** [comment]
I spent a couple of hours on this without much progress. I think constructing the `PlatformInfo` is relatively straightforward.

However, what I'm struggling with is how to compare that against the `target_compatible_with` attribute. How do I access the configured attributes? Can I even get access to the configured attributes before the original toolchain error happens?

**Comment by gregestren** [comment]
This might call for some virtual pair programming. Could you share with me a branch with your code? I'd also need to look down at the code level to sort through this stuff.

**Comment by philsc** [comment]
Here are my hacks from last weekend: https://github.com/philsc/bazel/tree/unreviewed/objc_library

Specifically, I think this is the last place where we can evaluate incompatibility:
https://github.com/philsc/bazel/blob/8f78bbf41a5a7eb17141f36d26b647350bcc425c/src/main/java/com/google/devtools/build/lib/skyframe/ConfiguredTargetFunction.java#L326

(Or we could add something inside of `computeDependencies()`, but I'm pretty sure we never get after `computeDependencies()` because in that callchain is where the toolchain error happens)

i had trouble going from `target` (and the other available information) to something that lets me access the configured attributes.

**Comment by glerchundi** [comment]
Hi 👋! Running into this issue as well, is there any progress on this? Thanks!!

**Comment by gregestren** [comment]
@philsc were you waiting for followup from me?

**Comment by philsc** [comment]
> @philsc were you waiting for followup from me?

Not really, no. I've just been extremely lazy in regards to bazel these past few months :P
I can take another look this weekend.

**Comment by philsc** [comment]
Apologies for the delay. Last weekend was too busy, but this weekend I'm planning to look at this again.

**Comment by glerchundi** [comment]
Thanks! 🙏

**Comment by philsc** [comment]
I'm starting to remember why I didn't solve this the last time I looked into it.
I need to understand this part of the code better in order to make changes here. The guidance @gregestren has provided helps, but I still need to understand this code better. It's currently still a bit "magic".
In other words, I feel like I'm making progress, but it's slow going. At the moment it's an archeology trip.

**Comment by philsc** [comment]
I feel like I'm close.

@gregestren , I need a way to convert the `Label`s from `ConfiguredAttributeMapper` to either `ConstraintValueInfo` instances or `<? extends ProviderCollection>` instances. See the `MAGIC` note in my current hacky commit:
https://github.com/philsc/bazel/blob/ab10d8621c79b0c22329af3674a1de2566392129/src/main/java/com/google/devtools/build/lib/skyframe/ConfiguredTargetFunction.java#L336-L361

In `RuleContextConstraintSemantics` we could use `ruleContext.getPrerequisites("target_compatible_with")` to get a provider collection, but I don't think it's an option here.

Any thoughts you have would be much appreciated!

**Comment by philsc** [comment]
Actually, it looks like the following code works to determine the equivalent `TransitiveInfoCollection` targets:
https://github.com/philsc/bazel/blob/66091e57c46540400d354172adf1c01bb27de994/src/main/java/com/google/devtools/build/lib/skyframe/ConfiguredTargetFunction.java#L346-L401

So I think we have a way to detect direct incompatibility. Now we just need to call `RuleContextConstraintSemantics.createIncompatibleConfiguredTarget()` but without a `ruleContext`. Sounds plausible/doable, but I haven't looked at that yet.

**Comment by gregestren** [comment]
That all looks about right to me.

I'd even bypass `resolveConfiguredTargetDependencies`, which has a bunch of extra logic you don't really need for this, and try going directly for calling another `ConfiguredTargetFunction` for the dep. For example, this pattern:

https://github.com/bazelbuild/bazel/blob/b3aacbbcf9c499a0493908ee14dce66a9834a52b/src/main/java/com/google/devtools/build/lib/skyframe/ConfiguredTargetFunction.java#L848-L854

In short, construct a `ConfiguredTargetKey` with the current configuration and the new label, call `env.getValue()` on it, and the result should directly give you a `ConfiguredTargetAndData` from which you can get a `TransitiveInfoCollection`.

I *think* that would produce slightly less code (although I could be wrong).

**Comment by gregestren** [comment]
Here's another example of translating `ConfiguredTargetKey`s into `ConfiguredTargetAndData`s:

https://github.com/bazelbuild/bazel/blob/e362fc9370f018bdf55ec7158ca420e3e6c18e25/src/main/java/com/google/devtools/build/lib/skyframe/PlatformLookupUtil.java#L68

https://github.com/bazelbuild/bazel/blob/e362fc9370f018bdf55ec7158ca420e3e6c18e25/src/main/java/com/google/devtools/build/lib/skyframe/PlatformLookupUtil.java#L142-L147

That example further casts into `PlatformInfo`, which is more work than you need to do. But the core logic is following the same sequence.

**Comment by philsc** [comment]
Sorry for dropping the ball on this ticket. It's been a busy few months. I'm resuming work on this at the end of August.

If anyone else wants to take it over in the meantime, I can totally understand that.

**Comment by philsc** [comment]
I created a draft PR after rebasing on latest master. Still need to actually incorporate Greg's feedback.

**Comment by philsc** [comment]
Note to self: test how this refactor behaves with `select()` against the toolchain:
```python
config_setting(
    name = "compiler_gcc",
    flag_values = {
        "@bazel_tools//tools/cpp:compiler": "gcc",
    },
)

cc_library(
    name = "gcc-only-lib",
    target_compatible_with = select({
        ":compiler_gcc": [],
        "//conditions:default": ["@platforms//:incompatible"],
    }),
    ...
)
```

**Comment by gregestren** [comment]
Added @philsc as co-assignee since https://github.com/bazelbuild/bazel/pull/14096 should cover this.

**Comment by pcjanzen** [comment]
Note that the refactoring done in 72787a1 doesn't solve this problem in all cases. Consider
```
==> BUILD <==
objc_library(
    name = 'ble',
    target_compatible_with = [
        "@platforms//os:macos",
    ],
)

platform(
    name = "macos",
    constraint_values = [
        "@platforms//os:macos",
    ],
)

==> platform_mappings <==
platforms:
  //:macos
    --cpu=darwin_x86_64

flags:
  --cpu=darwin_x86_64
    //:macos
```

In this case, objc_library's self-transition on --cpu leads to a transition on --platforms, and then ble's target platform is actually target_compatible_with macos again (regardless of the value for --platforms passed on the command line). So a toolchain is once again required.

**Comment by philsc** [comment]
Ah dang. Looks like the `platform_mappings` file is still triggering this issue. I will think about possible solutions.

**Comment by philsc** [comment]
Filed #16099 as a follow-up.
